drivertest/commontest/common.go

```diff
@@ -0,0 +1,284 @@
+// Copyright 2021 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package commontest
+
+import (
+	"database/sql"
+	"flag"
+	"io"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/cockroachdb/copyist"
+	"github.com/cockroachdb/copyist/drivertest/dockerdb"
+	"github.com/jmoiron/sqlx"
+	"github.com/stretchr/testify/require"
+)
+
+// DataTypes contains many interesting data types that can be returned by SQL
+// drivers.
+type DataTypes struct {
+	Int              int
+	Str, Dec, FltArr string
+	TimeZ, Time      time.Time
+	Bool             bool
+	Bytes            []byte
+	Flt              float64
+	Uuid             []byte
+}
+
+// resetScript is a generic SQL script that resets the database to a clean
+// state and creates some simple fixtures for common tests to use.
+const resetScript = `
+DROP TABLE IF EXISTS customers;
+CREATE TABLE customers (id INT PRIMARY KEY, name TEXT);
+INSERT INTO customers VALUES (1, 'Andy'), (2, 'Jay'), (3, 'Darin');
+
+DROP TABLE IF EXISTS datatypes;
+`
+
+// RunTests is called by other driver-specific test packages (like pgxtest and
+// pqtest) in order to set up the test environment and then run all tests. It
+// registers a copyist driver and starts up a SQL docker instance if in
+// recording mode. It then runs all tests by calling testing.M.Run(), and
+// finally exits the process when complete.
+func RunTests(m *testing.M, driverName, dataSourceName, dockerArgs string) {
+	flag.Parse()
+
+	copyist.Register(driverName, func() {
+		db, err := sql.Open(driverName, dataSourceName)
+		if err != nil {
+			panic(err)
+		}
+		defer db.Close()
+		if _, err := db.Exec(resetScript); err != nil {
+			panic(err)
+		}
+	})
+
+	// If in recording mode, then run database in docker container until test is
+	// complete.
+	var closer io.Closer
+	if copyist.IsRecording() {
+		closer = dockerdb.Start(dockerArgs, driverName, dataSourceName)
+	}
+
+	code := m.Run()
+
+	// Close the docker container before calling os.Exit; defers don't get
+	// called in that case.
+	if closer != nil {
+		closer.Close()
+	}
+
+	os.Exit(code)
+}
+
+// TestQuery fetches a single customer.
+func TestQuery(t *testing.T, driverName, dataSourceName string) {
```

Comment:
I was confused when initially reading this to see a `Test*` function in a non `*_test.go` file. Makes sense to me now what is going on. Not sure if there is anything to be done about that. Perhaps prefix this functions with some other word such as `DoTestQuery` or `RunTestQuery`, though that is a bit ugly. Fine to leave as-is.

Reply:
I switched to use `RunTestXXX`.